tiny_cnn/layers/average_pooling_layer.h

```diff
@@ -108,6 +108,85 @@ class average_pooling_layer : public partial_connected_layer<Activation> {
 
     std::string layer_type() const override { return "ave-pool"; }
 
+    void forward_propagation(cnn_size_t index,
+                             const std::vector<vec_t*>& in_data,
+                             std::vector<vec_t*>& out_data) override {
+        const vec_t& in  = *in_data[0];
+        const vec_t& w   = *in_data[1];
+        const vec_t& b   = *in_data[2];
+        vec_t&       out = *out_data[0];
+        vec_t&       a   = *out_data[1];
+
+        CNN_UNREFERENCED_PARAMETER(index);
+
+        auto oarea = out_.area();
+        size_t idx = 0;
+        for (size_t d=0; d<out_.depth_; ++d) {
+            float_t weight = w[d] * scale_factor_;
+            float_t bias = b[d];
+            for (size_t i=0; i<oarea; ++i, ++idx) {
+                const wi_connections& connections = out2wi_[idx];
+                float_t value = float_t(0);
+                for (auto connection : connections)// 13.1%
+                    value += in[connection.second]; // 3.2%
+                value *= weight;
+                value += bias;
+                a[idx] = value;
+            }
+        }
+
+        assert(out.size() == out2wi_.size());
+        for_i(parallelize_, out2wi_.size(), [&](int i) {
+            out[i] = h_.f(a, i);
+        });
+    }
+
+    void back_propagation(cnn_size_t                 index,
+                          const std::vector<vec_t*>& in_data,
+                          const std::vector<vec_t*>& out_data,
+                          std::vector<vec_t*>&       out_grad,
+                          std::vector<vec_t*>&       in_grad) override {
+        const vec_t& prev_out = *in_data[0];
+        const vec_t& w  = *in_data[1];
+        vec_t&       dW = *in_grad[1];
+        vec_t&       db = *in_grad[2];
+        vec_t&       prev_delta = *in_grad[0];
+        vec_t&       curr_delta = *out_grad[0];
+
+        CNN_UNREFERENCED_PARAMETER(index);
+
+        this->backward_activation(*out_grad[0], *out_data[0], curr_delta);
+
+        auto inarea = in_.area();
+        size_t idx = 0;
+        for (size_t i=0; i<in_.depth_; ++i) {
+            float_t weight = w[i] * scale_factor_;
+            for (size_t j=0; j<inarea; ++j, ++idx) {
```

Comment:
@edgarriba 
> you need to declare `scale_factor_` somewhere accessible for the class

The member is in its base class. `partial_connected_layer::scale_factor_`.

I copied `partial_connected_layer::forward_propagation` and `partial_connected_layer::back_propagation` methods into `average_pooling_layer` class for optimization. Speed-up factor from doing so is not very remarkable and some people may not disagree to this change as the change duplicates some code...